### PR TITLE
Add a Test Plan for OSPF Tests

### PR DESCRIPTION
**Pull request by asraza07**
### Description of PR
This PR contains a test plan for OSPF. The test plan comprises of two test cases:
1. Test OSPF Neighborship Status
2. Test Dynamic Routing Change

### Type of change
Added a new test plan to run OSPF along with OSPF test case code.

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205
- [x] 202305

#### What is the motivation for this PR?
Currently there is no existing test plan for running OSPF in SONiC. This test plan helps the user configure OSPF in the existing testbed and framework.

#### How did you do it?
The test cases disable the existing BGP configurations, enable the FRR container's OSPF feature, and establish OSPF neighborship using existing neighbor IP and DUT port IP configurations. Once these initial configurations are done, the planned tests and checks are carried out.

#### How did you verify/test it?
Run the test case on any topology

#### Any platform specific information?
It will run on all topologies where neighboring devices are SONiC devices.

#### Supported testbed topology if it's a new test case?
- [x] T0 (all variations)
- [x] T1 (all variations)

### CODE REVIEW

File: docs/testplan/OSPF-test-plan.md
Comment (yxieca) [inline docs/testplan/OSPF-test-plan.md:74]: Suggest verify the routes received was the same as in BGP configuration.

File: docs/testplan/OSPF-test-plan.md
Comment (yxieca) [inline docs/testplan/OSPF-test-plan.md:73]: Suggest make the OSPF configuration a plugin that can be extended to cEOS neighbors.

Make OSPF configuration and tear down plugin invokable from outside the test. and it should be smart enough to detect the config is already in OSPF therefore skip config/teardown.

Making config/teardown independent enough so that it can be used as a test-suite in the future: e.g. config once, run the test in this plan, run other existing IO tests to verify data plane, then tear down.

File: docs/testplan/OSPF-test-plan.md
Comment (yxieca) [inline docs/testplan/OSPF-test-plan.md:84]: Suggest also test receiving newly learnt routes from neighbors.

File: docs/testplan/OSPF-test-plan.md
Comment (asraza07) [reply docs/testplan/OSPF-test-plan.md]: Incorporated as per suggestion.

File: docs/testplan/OSPF-test-plan.md
Comment (asraza07) [reply docs/testplan/OSPF-test-plan.md]: Incorporated in test plan as per suggestion. We will implement the plugins in our test case code.

File: docs/testplan/OSPF-test-plan.md
Comment (asraza07) [reply docs/testplan/OSPF-test-plan.md]: Incorporated as per suggestion.

**Comment by asraza07** [comment]
@yxieca Kindly help review/merge this test plan after approval. Thanks.

**Comment by yxieca** [comment]
> @yxieca The above suggestions have been incorporated into the test plan. Kindly help review/merge this test plan after approval. Thanks.

@asraza07 thanks for updating the test plan. it looks good to me now. However, the test doesn't match the test plan in terms of the newly added scenarios. I suggest remove the test code from this PR and come back with test case PR matches the test plan.

**Comment by asraza07** [comment]
> > @yxieca The above suggestions have been incorporated into the test plan. Kindly help review/merge this test plan after approval. Thanks.
> 
> @asraza07 thanks for updating the test plan. it looks good to me now. However, the test doesn't match the test plan in terms of the newly added scenarios. I suggest remove the test code from this PR and come back with test case PR matches the test plan.

@yxieca I have removed the test case code from this PR. Please help review/merge. Thanks.

File: docs/testplan/OSPF-test-plan.md
Comment (asraza07) [reply docs/testplan/OSPF-test-plan.md]: @yxieca , I was looking into the code implementation. Currently, in the standard BGP configuration of the T0 topology, there are dozens of BGP routes available in the DUT which have been shared via neighboring (T1) devices (these routes originate from ExaBGP). These are BGP routes shared via BGP protocol from PTF to neighbor (T1) devices as BGP neighborship is established between PTF and T1 devices. I have attached a screenshot of the BGP configuration of the T1 device below for reference. In order for these routes to be passed on to the DUT via new OSPF configuration, the T1 devices will need to be running OSPF and BGP protocols simultaneously, as these routes from PTF will disappear if we disable BGP. As per my understanding, running two routing protocols together is not something conventionally done and can lead to issues. Can you please suggest for both BGP/OSPF work together. Thanks.

Screenshot from 2024-01-12 16-06-28

**Comment by ridahanif96** [comment]
@yxieca can you please help merge this PR.

**Review by yxieca — APPROVED** [review]

**Pull request merged**
